cmd/ocp-build-data-enforcer/main.go

```diff
@@ -0,0 +1,304 @@
+	"github.com/pmezard/go-difflib/difflib"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/github"
+)
+
+type options struct {
+	ocpBuildDataRepoDir string
+	majorMinor          majorMinor
+}
+
+func gatherOptions() *options {
+	o := &options{}
+	flag.StringVar(&o.ocpBuildDataRepoDir, "ocp-build-data-repo-dir", "../ocp-build-data", "The directory in which the ocp-build-data reposity is")
+	flag.StringVar(&o.majorMinor.minor, "minor", "6", "The minor version to target")
+	flag.Parse()
+	return o
+}
+func main() {
+	opts := gatherOptions()
+	opts.majorMinor.major = "4"
+
+	configsUnverified, err := gatherAllOCPImageConfigs(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to gather all ocp image configs")
+	}
+
+	streamMap, err := readStreamMap(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to read streamMap")
+	}
+
+	groupYAML, err := readGroupYAML(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Faild to read groupYAML")
+	}
+
+	var errs []error
+	var configs []ocpImageConfig
+	for _, cfg := range configsUnverified {
+		if err := cfg.validate(); err != nil {
+			errs = append(errs, fmt.Errorf("error validating %s: %w", cfg.SourceFileName, err))
+			continue
+		}
+		if err := dereferenceConfig(&cfg, streamMap, groupYAML); err != nil {
+			errs = append(errs, fmt.Errorf("failed dereferencing config for %s: %w", cfg.SourceFileName, err))
+			continue
+		}
+		configs = append(configs, cfg)
+	}
+
+	errGroup := &errgroup.Group{}
+	for idx := range configs {
+		idx := idx
+		errGroup.Go(func() error {
+			processDockerfile(configs[idx])
+			return nil
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		logrus.WithError(err).Fatal("Processing failed")
+	}
+
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		for _, err := range err.Errors() {
+			logrus.WithError(err).Error("Encountered error")
+		}
+		logrus.Fatal("Encountered errors")
+	}
+	logrus.Infof("Processed %d configs", len(configs))
+}
+
+// TODO (alvaroaleman): This is incomplete
+func dereferenceConfig(config *ocpImageConfig, streamMap streamMap, groupYAML groupYAML) error {
+	if replacement, hasReplacement := streamMap[config.From.Stream]; hasReplacement {
+		config.From.Stream = replacement.UpstreamImage
+	}
+	for blder := range config.From.Builder {
+		if replacement, hasReplacement := streamMap[config.From.Builder[blder].Stream]; hasReplacement {
+			config.From.Builder[blder].Stream = replacement.UpstreamImage
+		}
+	}
+	if config.Content.Source.Alias != "" {
+		if _, hasReplacement := groupYAML.Sources[config.Content.Source.Alias]; !hasReplacement {
+			return fmt.Errorf("groups.yaml has no replacement for alias %s", config.Content.Source.Alias)
+		}
+		config.Content.Source.Git = &ocpImageConfigSourceGit{}
```

Comment:
Why this line and the following? I don't understand this code

Reply:
The field is a pointer but I don't want to create a pointer to `groupYAML.Sources[config.Content.Source.Alias]` so I first set it to an empty object, then set the value of that empty object to `groupYAML.Sources[config.Content.Source.Alias]`

Reply:
Why is the field a pointer?

Reply:
To be able to tell if it was set in the yaml